### PR TITLE
Fix place order and more

### PR DESCRIPTION
**Pull request by bdowling**
Ok, let me straight-up say I know I'm going to get some flack for this PR.  I'm not submitting it so much to merge, but more so that I can let others know I've been working on these things.  I have a desire to get this module working for my needs, so I'm making changes at a bit of a rapid pace.

This makes it harder to submit more discrete PRs with small changes/fixes as is generally preferred.  SO I understand if this is too much at once.  Please take it with a grain of salt.  I did separate my commits a bit, but for the sake of just putting this out there, I'm submitting this as one PR.

A few of my changes may be breaking the prior API exposed, so please look at this with that in mind.

In particular, I noted the issues with place_order and not being able to reliably do LIMIT and STOP LIMIT/MARKET orders.  I tried to fix this up as I thought appropriate, and also added some Enums to follow the pattern to make the various reserved words more "structured" in passing them around.

I'm am also interested if someone has discussed an approach we should take to start using requests-mock -- this would allow us to craft more testing around these sort of blanket changes.

### CODE REVIEW

File: Robinhood/Robinhood.py [hunk 1]
```diff
@@ -165,53 +195,91 @@ def investment_profile(self):
         data = res.json()
         return data
 
-    def instruments(self, stock):
+    def instruments(self, query=None, symbol=None, instrumentid=None):
```
Comment (lockefox) [inline Robinhood/Robinhood.py]: though `symbol` is the right word in this case, `stock` is the uniform name in the API

Also, `instrument_id` would be my nit-pick for naming

File: Robinhood/Robinhood.py [hunk 2]
```diff
@@ -274,6 +343,37 @@ def get_quote(self, stock=''):
...
+            fields.append('symbol')
+
+        res = {}
+        for quote in data['results']:
+            print(type(quote))
+            if not isinstance(quote,dict):
+                raise Warning("Returned Quote was not a dict")
+
+            # XXX what if symbol is non-unique should we use symbol as key here?
+            s = quote['symbol']
```
Comment (lockefox) [inline Robinhood/Robinhood.py]: expand `s` to `symbol`

File: Robinhood/Robinhood.py [hunk 3]
```diff
@@ -700,38 +820,61 @@ def place_order(
...
 
         Returns:
             (:obj:`requests.request`): result from `orders` put command
 
         """
+
         if isinstance(transaction, str):
-            transaction = Transaction(transaction)
-        if not bid_price:
+            transaction = Transaction(transaction.lower())
```
Comment (lockefox) [inline Robinhood/Robinhood.py]: checking for `str` seems clumsy, just try to set directly rather than if check?

File: Robinhood/Robinhood.py [hunk 4]
```diff
@@ -778,3 +920,45 @@ def place_sell_order(
...
+        convenience function to cancel all orders, optionally only for a given instrument
+
+        Args:
+            instrument (string): RH instrument URL to restrict results
+        Returns:
+            (:obj:`dict`): containing keys 'cancelled' or 'error' with the list of order ids
+        """
+        orders = self.order_history(instrument=instrument)
+        res = {'cancelled':[], 'error':[]}
+        stoppedOrderStates = (OrderState.FILLED, OrderState.CANCELLED, OrderState.FAILED, OrderState.REJECTED)
```
Comment (lockefox) [inline Robinhood/Robinhood.py]: move `stopped_order_states` up to class-global and make it `_stopped_order_states` as a private collection

File: Robinhood/Robinhood.py (hunk 1 above)
Comment (bdowling) [reply Robinhood/Robinhood.py]: fwiw, previously this function was not searching for a stock, it was using the query param to search by name (e.g. company name or symbols), so I felt making it more explicit was better here.

File: Robinhood/Robinhood.py (hunk 4 above)
Comment (bdowling) [reply Robinhood/Robinhood.py]: ah, yes, I forgot to move this and make it more generally available.  Given all the States RH api can return I figured we'd want these access to these groupings.  I was thinking convenience  methods could work `isStoppedOrderStatus()`.  Could be added to the Enum and/or through subclassing and using isinstance(state, OrderStopped).  importing the Enums outside the library also feels clumsy right now, I might be missing something tho..
```
from Robinhood.Robinhood import Transaction
```
Should they be within class Robinhood instead?

Comment (bdowling) [reply Robinhood/Robinhood.py]: I added a commit to move the Enums and test state.  I think this is a much cleaner approach.

File: Robinhood/Robinhood.py [hunk 5]
```diff
@@ -776,5 +923,47 @@ def place_sell_order(
             (:obj:`requests.request`): result from `orders` put command
 
         """
-        transaction = Transaction.SELL
+        transaction = self.Transaction.SELL
         return self.place_order(instrument, quantity, bid_price, transaction)
+
```
Comment (liquidgenius) [inline Robinhood/Robinhood.py]:
```
    def place_stop_limit_order(
            self,
            instrument,
            quantity,
            bid_price=None,
            stop_price=0.0,
            trigger = Trigger.STOP,
            order = Order.MARKET,
            time_in_force = TimeForce.GFD,
    ):
        print(trigger)
        print(order)

        """wrapper for placing stop limit orders

        Args:
            instrument (dict): the RH URL and symbol in dict for the instrument to be traded
            quantity (int): quantity of stocks in order
            stop_price (float): the price at which the stop will trigger (?)
            (see coding around line #860-61 regarding bid_price)

        Returns:
            (:obj:`requests.request`): result from `orders` put command

        """
        transaction = Transaction.SELL
        return self.place_order(instrument, quantity, transaction, bid_price, stop_price, trigger, order, time_in_force)
```

File: Robinhood/Robinhood.py [hunk 6]
```diff
@@ -700,38 +829,56 @@ def place_order(
...
+        if order == self.Order.LIMIT and bid_price is None:
+            raise ValueError("Order.LIMIT without bid_price")
+
+        if bid_price is not None and not order == self.Order.LIMIT:
+            raise ValueError("bid_price without Order.LIMIT")
+
+        # If Order.MARKET the we still need to pass RH a price, RH places market orders
+        # automatically as limit orders with a 5% buffer
+        if not bid_price and order == self.Order.MARKET:
+            quote = self.quote_data(instrument['symbol'])['results'][0]
```
Comment (liquidgenius) [inline Robinhood/Robinhood.py]: bid_price = self.quote_data(instrument['symbol'])['results'][0]['bid_price']

File: Robinhood/Robinhood.py (hunk 5 above)
Comment (liquidgenius) [reply Robinhood/Robinhood.py]: Tested, working in live environment.

File: Robinhood/Robinhood.py [hunk 7]
```diff
@@ -37,10 +25,13 @@ class Robinhood:
...
+        "instrumentid": "https://api.robinhood.com/instruments/{instrumentid}/",
+        "instrumentsplits": "https://api.robinhood.com/instruments/{instrumentid}/splits/",
         "margin_upgrades": "https://api.robinhood.com/margin/upgrades/",
         "markets": "https://api.robinhood.com/markets/",
         "notifications": "https://api.robinhood.com/notifications/",
         "orders": "https://api.robinhood.com/orders/",
+        "cancel": "https://api.robinhood.com/orders/{oid}/cancel/",
         "password_reset": "https://api.robinhood.com/password_reset/request/",
         "portfolios": "https://api.robinhood.com/portfolios/",
         "positions": "https://api.robinhood.com/positions/",
```
Comment (liquidgenius) [inline Robinhood/Robinhood.py]: `        "sp500_up": "https://api.robinhood.com/midlands/movers/sp500/?direction=up",
        "sp500_dn": "https://api.robinhood.com/midlands/movers/sp500/?direction=dn",`

**Comment by Jamonek** [comment]
We have a few conflicts to work on.

**Comment by Jamonek** [comment]
Any updates?

**Comment by liquidgenius** [comment]
I added some other order types and automatic generation and updating of
local sqlite database that matches the Robinhood datafeeds, but moved
recently and I'm not good with Github check-in and check-out process so I
haven't uploaded the changes.

-John

**Comment by bdowling** [comment]
I was going to work this week and try to resolve the conflicts and make this cleanly merge upstream, but I have run into a number of competing changes / conflicts.

The changes merged in #55 in particular have made a mess of conflicts.  I'm not all that inclined to resolve these because I do not see why a competing order system needed to be merged without comment.  

I have submitted additional commits to this branch and believe I have improved the order functions and usage of the Enums that have been added do makes the API cleaner.  

As I've mentioned before, having so many support functions (wrappers) really should be broken out into their own support library, and many of them should be dynamic/generated functions given the boilerplate repetition.

Given the competing interests, I may just need to continue working in my fork to satisfy my own evolution of this library.  I had hoped to not do that, but that may be for the best.  

I welcome comments if others are interested in the work I've done here.

**Comment by ciresnave** [comment]
I would be interested in working with you and anyone else interested to bring this API to where it should be at. It sounds like you and I agree that helper functions need to be separated out of this. From other comments I've seen you make I think you would probably agree that this API needs to be limited to the actual RobinHood interface and should support all of it's direct features without any extra frills...correct? I started to throw together a few pages in the wiki for this repository in hopes we could all start collaborating and properly planning the future of this API. Let's work together rather than competing.

**Comment by bdowling** [comment]
Ok, I lied.  I made an attempt at resolving all the conflicts and rebased this branch on master, it was a bit of a mess given some of the divergence, but I think I got it mostly right in this branch:

https://github.com/bdowling/Robinhood/tree/fix_place_order_and_more_merge

(Of course it would be nice if we had a good regression suite to validate such heavy changes!!)

Please feel free to take a look, if there is no dissent, I will squash my commits and force push it into this PR.  I just didn't want to rewrite history in this branch as I know a few others have forked and made some of their own improvements off this in-progress PR.

fyi... @TomFaulkner, @liquidgenius et al

**Comment by liquidgenius** [comment]
No dissent here. I added features that wouldn’t really be considered part
of the core. IE, local SQLite database, etc.

Better if I keep them separate.

-JZ
@liquidgenius

**Comment by Jamonek** [comment]
Hey everyone,

I’ll modify the process for merging work into the master branch. Starting with #44 we’ll discuss the changes made before merging unless it is a hotfix. 

Best,

Jamone

**Comment by bdowling** [comment]
@liquidgenius Yes, the local sqlite database is a good feature, perhaps not part of the core, but the core should perhaps provide support for such a Caching layer, imho.  Maybe it could be done through inheritance.  It's a good feature that I was also considering to have to improve usage and lighten the load on RH API.

**Comment by lockefox** [comment]
Managing an internal caching layer is some significant feature creep and would require serious test coverage to make sure we don't screw users with stale data.

If you must cache, [request-cache](https://pypi.python.org/pypi/requests-cache) might be the tool.  [pandas-datareader](https://pypi.python.org/pypi/pandas-datareader) gives hooks to allow library tune/swap from outside

**Comment by TomFaulkner** [comment]
@bdowling I'm okay with that, I haven't made any unpushed changes.

**Pull request closed without merging**